jdaviz/configs/default/plugins/subset_plugin/subset_plugin.py

```diff
@@ -655,3 +655,23 @@ def _set_value_in_subset_definition(self, index, name, desired_key, new_value):
             if self.subset_definitions[index][i]['name'] == name:
                 self.subset_definitions[index][i]['value'] = new_value
                 break
+
+    def _import_spectral_regions(self, spec_region, mode=NewMode):
+        """
+        Method for importing a SpectralRegion object or list of SpectralRegion objects.
+
+        Parameters
+        ----------
+        spec_region : ~`specutils.SpectralRegion` object or list
+            The object that contains bounds for the spectral region or regions
+        mode : AndMode, AndNotMode, OrMode, ReplaceMode, XorMode, or NewMode
```

Comment:
Do we want to allow `mode` to also be a list of the same length as `spec_region` list, to allow building more complicated composite region with one call?

Reply:
If `spec_region` is list but `mode` is scalar, then we broadcast like you are doing now. if `spec_region` is scalar but `mode` is a list, we could either grab only the first element or throw exception (I prefer throwing exception).

Reply:
Spectral subsets can logically be represented as either one single subregion or as multiple subregions combined by `OrMode`. I don't think there's a great reason to make mode a list unless we want to allow complex composite spectral subsets (i.e. composite subsets that were created with `Xor`, `And`, and `AndNot`) to be loaded in without forcing them to be simplified first.

Reply:
Do we disallow complicated spectral subsets? I thought you can use any of the exposed subset operators in the dropdown.

Reply:
We do not disallow them but all spectral subsets can be represented with the `Or` operator. For example, if you create a spectral subset from 1 to 5 and use `AndNotMode` on 3 to 4, you can represent that subset with 1 to 3 `OrMode` 4 to 5.

Reply:
So it would be "(1 and 2) or (5)"? So technically you still need to allow more than one kind of operator for more complicated stuff?

Reply:
I'm not sure I understand. In the CubevizExample notebook, create a composite spectral subset with all the operators and then run `cubeviz.app.get_subsets()`. The returned value from that will explain what I'm trying to say here.

Reply:
I tried it in Specviz but I got a traceback, so not sure what I am supposed to see.

```
----> 1 specviz.app.get_subsets()

File jdaviz/app.py:947, in Application.get_subsets(self, subset_name, spectral_only, spatial_only, object_only, simplify_spectral, use_display_units, include_sky_region)
    942 label = subset.label
    944 if isinstance(subset.subset_state, CompositeSubsetState):
    945     # Region composed of multiple ROI or Range subset
    946     # objects that must be traversed
--> 947     subset_region = self.get_sub_regions(subset.subset_state,
    948                                          simplify_spectral, use_display_units,
    949                                          get_sky_regions=include_sky_region)
    951 elif isinstance(subset.subset_state, RoiSubsetState):
    952     subset_region = self._get_roi_subset_definition(subset.subset_state,
    953                                                     to_sky=include_sky_region)

File jdaviz/app.py:1224, in Application.get_sub_regions(self, subset_state, simplify_spectral, use_display_units, get_sky_regions)
   1222     new_region = temp_region.invert(one.lower, one.upper)
   1223 else:
-> 1224     new_region = new_region + temp_region.invert(one.lower, one.upper)
   1225 # This adds the edge regions that are otherwise not included
   1226 if not (one.lower == temp_region.lower and one.upper == temp_region.upper):

File specutils/spectra/spectral_region.py:376, in SpectralRegion.invert(self, lower_bound, upper_bound)
    370 newlist = newlist[1:-1]
    372 #
    373 # Now create new Spectrum region
    374 #
--> 376 return SpectralRegion([(x, y) for x, y in zip(newlist[0::2], newlist[1::2])])

File specutils/spectra/spectral_region.py:56, in SpectralRegion.__init__(self, *args)
     52     raise TypeError(f'SpectralRegion initializer takes 1 or 2'
     53                     f'positional arguments but {len(args)} were given')
     55 #  Check validity of the input sub regions.
---> 56 self._valid()
     58 # The sub-regions are to always be ordered based on the lower bound.
     59 self._reorder()

File specutils/spectra/spectral_region.py:221, in SpectralRegion._valid(self)
    219 def _valid(self):
--> 221     bound_unit = self._subregions[0][0].unit
    222     for x in self._subregions:
    223         if x[0].unit != bound_unit or x[1].unit != bound_unit:

IndexError: list index out of range
```

![Screenshot 2024-07-10 155502](https://github.com/spacetelescope/jdaviz/assets/2090236/d4504864-16c7-4bac-bf02-b93efa7ae0ee)
![Screenshot 2024-07-10 155515](https://github.com/spacetelescope/jdaviz/assets/2090236/5d7c554c-040e-4485-8446-d31a0622ae64)
![Screenshot 2024-07-10 155527](https://github.com/spacetelescope/jdaviz/assets/2090236/744985c6-a624-4f88-8fae-f5d0a604f8e1)
![Screenshot 2024-07-10 155539](https://github.com/spacetelescope/jdaviz/assets/2090236/44438fd3-0fbf-4a76-8467-be3cf3d05859)

Reply:
OK. I see your point that `app.get_subsets()` simplifies the composite subsets and returns only the simplified version. But at the same time, it also optionally allows `simplify_spectral=False`. As a user, I would want to create composite subset in a way that makes sense to me scientifically, not necessarily the simplified way, thus this low-level API should not disallow such a thing?

Reply:
I added the ability to use a list. I am considering using `SUBSET_MODES` at the top of `subset_plugin.py` to convert strings ('replace', 'OrState', 'AndState', etc.) instead of relying on the actual mode type being imported from glue. Let me know if you think that is a good idea.

Reply:
https://github.com/spacetelescope/jdaviz/pull/3124